src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+function as_polycyclic_group(G::DirectProductOfGroups)
+   if [typeof(H)==PcGroup for H in G.L]==[true for i in 1:G.n]
+      return PcGroup(G.X)
+   else
+      throw(ArgumentError("The group is not a polycyclic group"))
+   end
+end
+
+"""
+    as_matrix_group(G::DirectProductOfGroups)
+If `G` is direct product of matrix groups over the ring `R` of dimension `n_1`, ... , `n_k` respectively, return `G` as matrix group over the ring `R` of dimension `n_1 + ... + n_k`.
+"""
+function as_matrix_group(G::DirectProductOfGroups)
+#   if S==MatrixGroup && T==MatrixGroup && GAP.Globals.FieldOfMatrixGroup(G.G1.X)==GAP.Globals.FieldOfMatrixGroup(G.G2.X)
+   if [typeof(H)==MatrixGroup for H in G.L]==[true for i in 1:G.n] && length(Set([base_ring(H) for H in G.L]))==1
+      return MatrixGroup(G.X)
+   else
+      throw(ArgumentError("The group is not a matrix group"))
+   end
+end
+
+"""
+    embedding(G::DirectProductOfGroups, j::Integer)
+Return the embedding of the `j`-th component of `G` into `G`, for `j` = 1,...,#factors of `G`.
+"""
+function embedding(G::DirectProductOfGroups, j::Base.Integer)
+   if !G.isfull
+      throw(ArgumentError("Embedding is not defined for proper subgroups of direct products"))
+   end
+   if j in 1:G.n
+      f=GAP.Globals.Embedding(G.X,j)
+      gr=G.L[j]
+      typeG=typeof(gr)
+   else
+      throw(ArgumentError("index not valid"))
+   end
+   return GAPGroupHomomorphism{typeG,DirectProductOfGroups}(gr,G,f)
+end
+
+"""
+    projection(G::DirectProductOfGroups, j::Integer)
+Return the projection of `G` into the `j`-th component of `G`, for `j` = 1,...,#factors of `G`.
+"""
+function projection(G::DirectProductOfGroups, j::Base.Integer)
+  # @assert W.isfull "Projection not defined for proper subgroups of wreath products"
+   f=GAP.Globals.Projection(G.Xfull,j)
+   if j in 1:G.n
+      H = G.L[j]
+      if !G.isfull
+         Gf = DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+         g = embedding(Gf,G)
+         p = g*GAPGroupHomomorphism{DirectProductOfGroups,typeof(H)}(Gf,H,f)
+      else
+         p = GAPGroupHomomorphism{DirectProductOfGroups,typeof(H)}(G,H,f)
+      end
+      return hom(G,H,x->p(x))
+   else
+      throw(ArgumentError("index not valid"))
+   end
+end
+
+function (G::DirectProductOfGroups)(V::Vector{GAPGroupElem})
+   Gfull=DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+   @assert length(V)==G.n "Vector of wrong length"
+   x = prod([embedding(Gfull,j)(V[j]) for j in 1:length(V)])
+   @assert x in G "Element not in the group"
+   return group_element(G,x.X)
+end
+
+function (G::DirectProductOfGroups)(V::GAPGroupElem...)
+   Gfull=DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+   @assert length(V)==G.n "Vector of wrong length"
+   x = prod([embedding(Gfull,j)(V[j]) for j in 1:length(V)])
+   @assert x in G "Element not in the group"
+   return group_element(G,x.X)
+end
+
+# start part on subgroups
+function _as_subgroup_bare(G::DirectProductOfGroups, H::GapObj)
+  t = H==G.X
```

Comment:
This check is potentially super expensive! It may even be undecidable if some of the factors are fp groups. Do we really need to do this? It will make working with subgroups basically unusable except in special cases (e.g. when all factors are permutation groups, or all are very small, or ...)

Reply:
It is not strictly necessary. The effect of this check is to assign the value `true` to the field `H.isfull` when the subgroup `H` coincide with the whole group `G`. We can delete this check so a not full group `H` will be always create, and delegate the job to decide whether `H` is full only to the function `write_as_full`.